// apollo-language-server/lib/engine/operations/validateOperations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_tag_1 = require("graphql-tag");
var ValidationErrorType;
(function (ValidationErrorType) {
    ValidationErrorType["WARNING"] = "WARNING";
    ValidationErrorType["FAILURE"] = "FAILURE";
    ValidationErrorType["INVALID"] = "INVALID";
})(ValidationErrorType = exports.ValidationErrorType || (exports.ValidationErrorType = {}));
var ValidationErrorCode;
(function (ValidationErrorCode) {
    ValidationErrorCode["NON_PARSEABLE_DOCUMENT"] = "NON_PARSEABLE_DOCUMENT";
    ValidationErrorCode["INVALID_OPERATION"] = "INVALID_OPERATION";
    ValidationErrorCode["DEPRECATED_FIELD"] = "DEPRECATED_FIELD";
})(ValidationErrorCode = exports.ValidationErrorCode || (exports.ValidationErrorCode = {}));
exports.VALIDATE_OPERATIONS = graphql_tag_1.default `
  mutation ValidateOperations(
    $id: ID!
    $operations: [OperationDocumentInput!]!
    $tag: String
    $gitContext: GitContextInput
  ) {
    service(id: $id) {
      validateOperations(
        tag: $tag
        operations: $operations
        gitContext: $gitContext
      ) {
        validationResults {
          type
          code
          description
          operation {
            name
          }
        }
      }
    }
  }
`;
